ADR/adr-0008-secret-mgmt-for-user-workloads.md

```diff
@@ -0,0 +1,105 @@
+# Secret Management For User Workloads
+
+Date: 2022-10-28
+
+## Status
+
+Accepted
+
+## Context
+
+* When user workloads are deployed to environments, the system should be able to provide a way to inject values that are specific to the environment. Currently, this is done through environment variables that are managed as overlays on the GitOps repository for the application. However, this method does not provide a good way to manage `Secret`. This ADR addresses the secret management of user workloads for different environments.
+* As a StoneSoup component, I expect to have the ability to securely upload a secret, associate it with the component, and propagate it to the target environment.
+
+## Decision
+
+### Architecture Overview
+#### Terminology
+
+- `UploadSecret`: A short-lived Kubernetes secret used to deliver confidential data to permanent storage and link it to the RemoteSecret CR.
+- `SecretData`: An object stored in permanent SecretStorage. Valid SecretData is always linked to a RemoteSecret CR.
+- `RemoteSecret`: A CRD that appears during upload and links `SecretData` + `DeploymentTarget(s)` + K8s `Secret`. `RemoteSecret` is linked to one (or zero) SecretData and manages its deleting/updating.
+- K8s `Secret`: What appears at the output and is used by consumers.
+- `SecretId`: A unique identifier of SecretData in permanent SecretStorage.
+- `SecretStorage`: A database eligible for storing `SecretData` (such as HashiCorp Vault, AWS Secret Manager).
+
+#### Architecture
+
+![](../diagrams/secret-mgmt.excalidraw.svg)
+
+The idea is to introduce a new CR called `RemoteSecret`, which is a Kubernetes representation of the `K8s Secret` that is stored in permanent storage aka SecretStorage. This CR contains a reference to the destination, either a Kubernetes namespace or an `Environment`. `UploadSecret` is used to perform an upload to the permanent storage and is represented as a regular Kubernetes secret with special labels and annotations recognized by the _SPI controller_. Different implementations of `SecretStorage` can be used, such as AWS Secret Manager or HashiCorp Vault.
+
+#### Example: If the destination is `Environment`
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: jdbc-connection-parameter
+  labels:
+    spi.appstudio.redhat.com/upload-secret: secret
+    spi.appstudio.redhat.com/target-environment: prod
```

Comment:
I think this needs one more label, to be the `name` of the `Secret` that is put in place in the deployment target as a result of the `RemoteSecret`.

To see why, consider this problem:

* User has two environments (stage and prod) and has a password secret that needs to be available in both.
* The user's app expects to mount the secret from a secret called `password-secret`.
* In the user's workspace, they create an **Upload Secret** with target-environment set to stage. The upload secret has name "password-secret-stage".
* This is reconciled and a RemoteSecret is created, also with metadata.name set to "stage-password" and also with spec.secret.name set to "password-secret-stage".

When the user's app is deployed in the stage environment, it will try to mount a secret called "password-secret", but the secret (in the stage environment, put there by way of the RemoteSecret) will be named "password-secret-stage".

```suggestion
  labels:
    spi.appstudio.redhat.com/upload-secret: secret
    spi.appstudio.redhat.com/target-environment: prod
    spi.appstudio.redhat.com/target-secret-name: password-secret
```